### PR TITLE
Linkcheck tweaks

### PR DESCRIPTION
**Pull request by electrofelix**
Switch to requests for better content encoding handling, fallback to GET requests on 403.

### CODE REVIEW

File: sphinx/builders/linkcheck.py [hunk 1]
```diff
@@ -36,28 +38,25 @@ class HTMLParseError(Exception):
...
+    pkg_resources.require(['requests[security]'])
+except pkg_resources.DistributionNotFound:
+    import ssl
+    if not getattr(ssl, 'HAS_SNI', False):
+        # don't complain on each url processed about the SSL issue
+        requests.packages.urllib3.disable_warnings(
+            requests.packages.urllib3.exceptions.InsecurePlatformWarning)
+        warnings.warn(
+            'Some links may return broken results due to being unable to '
+            'check the Server Name Indicate (SNI) in the returned SSL cert '
```
Comment (lehmannro) [inline sphinx/builders/linkcheck.py]: I think the proper term is _Server Name Indicat<b>ion.</b>_

File: doc/config.rst [hunk 2]
```diff
@@ -1905,6 +1905,15 @@ Options for the linkcheck builder
 
    .. versionadded:: 1.2
 
+.. confval:: linkcheck_anchors_ignore
```
Comment (lehmannro) [inline doc/config.rst]: Just a small nit:  I'd personally prefer this to be a separate PR.

It definitely needs a test.

File: sphinx/config.py [hunk 3]
```diff
@@ -257,6 +257,7 @@ class Config(object):
         linkcheck_timeout = (None, None, [int]),
         linkcheck_workers = (5, None),
         linkcheck_anchors = (True, None),
+        linkcheck_anchors_ignore = ([], None),
```
Comment (lehmannro) [inline sphinx/config.py]: Could we make this default to `['!']`, so we can delete this hard-coded path from the code?

File: sphinx/builders/linkcheck.py [hunk 4]
```diff
@@ -75,18 +74,18 @@ def handle_starttag(self, tag, attrs):
                 self.found = True
 
 
-def check_anchor(f, anchor):
-    """Reads HTML data from a filelike object 'f' searching for *anchor*.
+def check_anchor(response, anchor):
+    """Reads HTML data from a response object 'response' searching for *anchor*.
```
Comment (lehmannro) [inline sphinx/builders/linkcheck.py]: While we're here, make references to both arguments in the docstring uniform (`*response*`, or even better backticks.)

File: sphinx/builders/linkcheck.py [hunk 5]
```diff
@@ -137,10 +130,16 @@ def check_thread(self):
...
             kwargs['timeout'] = self.app.config.linkcheck_timeout
 
+        kwargs['allow_redirects'] = True
+
         def check_uri():
             # split off anchor
             if '#' in uri:
                 req_url, anchor = uri.split('#', 1)
+                for rex in self.skip_anchors:
+                    if rex.match(uri):
```
Comment (lehmannro) [inline sphinx/builders/linkcheck.py]: Shouldn't this be `rex.match(anchor)`?

File: sphinx/builders/linkcheck.py [hunk 6]
```diff
@@ -36,28 +38,25 @@ class HTMLParseError(Exception):
 from sphinx.util import encode_uri
 from sphinx.util.console import purple, red, darkgreen, darkgray, \
     darkred, turquoise
-from sphinx.util.pycompat import TextIOWrapper
```
Comment (lehmannro) [inline sphinx/builders/linkcheck.py:39]: I think this file was the sole user of `sphinx.util.pycompat.TextIOWrapper`, so if it's no longer referenced we can deprecate/remove it.

File: sphinx/builders/linkcheck.py [hunk 7]
```diff
@@ -36,28 +38,25 @@ class HTMLParseError(Exception):
...
+    if not getattr(ssl, 'HAS_SNI', False):
+        # don't complain on each url processed about the SSL issue
+        requests.packages.urllib3.disable_warnings(
+            requests.packages.urllib3.exceptions.InsecurePlatformWarning)
+        warnings.warn(
+            'Some links may return broken results due to being unable to '
+            'check the Server Name Indicate (SNI) in the returned SSL cert '
+            'against the hostname in the url requested. Recommended to '
+            'install "requests[security]" as a dependency or upgrade to '
+            'a python version with SNI support.'
```
Comment (lehmannro) [inline sphinx/builders/linkcheck.py]: We should say which that is.

File: sphinx/builders/linkcheck.py [hunk 8]
```diff
@@ -75,18 +74,18 @@ def handle_starttag(self, tag, attrs):
...
+    """Reads HTML data from a response object 'response' searching for *anchor*.
     Returns True if anchor was found, False otherwise.
     """
     parser = AnchorCheckParser(anchor)
     try:
         # Read file in chunks of 8192 bytes. If we find a matching anchor, we
         # break the loop early in hopes not to have to download the whole thing.
-        chunk = f.read(8192)
-        while chunk and not parser.found:
+        for chunk in response.iter_content(8192):
```
Comment (lehmannro) [inline sphinx/builders/linkcheck.py]: The default for `response.__iter__` is much smaller;  any reason we're not using their suggestion?

File: sphinx/builders/linkcheck.py (hunk 1 above)
Comment (electrofelix) [inline sphinx/builders/linkcheck.py]: Oops! I'll fix that up.

File: doc/config.rst (hunk 2 above)
Comment (electrofelix) [inline doc/config.rst]: Sure, wasn't clear to me exactly how to add a test for this, but I'll dig and and can discuss further in the separate PR anyway. It's a separate commit anyway so trivial to send it as a separate PR after the requests one.

File: sphinx/config.py (hunk 3 above)
Comment (electrofelix) [inline sphinx/config.py]: Will do, be either that or something equivalent.

File: sphinx/builders/linkcheck.py (hunk 4 above)
Comment (electrofelix) [inline sphinx/builders/linkcheck.py]: will do.

File: sphinx/builders/linkcheck.py (hunk 5 above)
Comment (electrofelix) [inline sphinx/builders/linkcheck.py]: I was thinking in terms of allowing matching against the full URI to allow skipping based on the site name, just in the off chance that the same anchor name was reused for multiple sites but some used JS preventing checking for it.

Perhaps that's something that can be enabled with a custom flag for the regex or by setting an additional option if needed and we can default to just matching against the anchor for now?

File: sphinx/builders/linkcheck.py (hunk 7 above)
Comment (electrofelix) [inline sphinx/builders/linkcheck.py]: Just specify the known versions of python 3 and 2.7.9+, and if someone is running an older one it depends on whether the distro patched the support in but ignore that for the message?

File: sphinx/builders/linkcheck.py (hunk 8 above)
Comment (electrofelix) [inline sphinx/builders/linkcheck.py]: Purely because I reused the selected size of 8192 from the existing code.

Quite happy to change it though.

File: sphinx/builders/linkcheck.py (hunk 6 above)
Comment (electrofelix) [inline sphinx/builders/linkcheck.py:39]: Appears to be used in `sphinx/pycode/__init__.py` as well.

**Comment by tk0miya** [comment]
LGTM!
I will merge this within a day (if no objections).
Could you add entry to CHANGES file?

**Comment by lehmannro** [comment]
No objections from my side.  Looking forward to the follow-up PR.

**Comment by tk0miya** [comment]
Thank you for contribution!

**Comment by tkelman** [comment]
What's the rough process / timeline for when this can be expected in a release?

edit: I found https://github.com/sphinx-doc/sphinx/commit/08cfcbc363d684510dff3b2ed628a88a0e4d729a, looks like it's scheduled for 1.5 then.

**Comment by tk0miya** [comment]
Not planned yet.
IMO, we have many improves in master branch. so it's time to major release.

**Comment by tkelman** [comment]
Thanks for the update. I realize all I actually need from this right now would be the change to make 403 also fall back to GET like 405 already does. That could probably be done on its own independent of the rest of the change, but that might be too much hassle to worry about.

**Comment by tk0miya** [comment]
@tkelman I can accept if you send pull requests to stable branch to support 403 fallbacks.
